@app/components/src/AvatarUpload.tsx

```diff
@@ -0,0 +1,208 @@
+import React, { useState, useEffect } from "react";
+import { Upload, Icon, message } from "antd";
+import { UploadChangeParam } from "antd/lib/upload";
+import { UploadFile, RcCustomRequestOptions } from "antd/lib/upload/interface";
+import axios from "axios";
+import {
+  useChangeAvatarMutation,
+  ProfileSettingsForm_UserFragment,
+} from "@app/graphql";
+import { ApolloError } from "apollo-client";
+
+export function slugify(string: string) {
+  const a =
+    "àáâäæãåāăąçćčđďèéêëēėęěğǵḧîïíīįìłḿñńǹňôöòóœøōõőṕŕřßśšşșťțûüùúūǘůűųẃẍÿýžźż·/_,:;";
+  const b =
+    "aaaaaaaaaacccddeeeeeeeegghiiiiiilmnnnnoooooooooprrsssssttuuuuuuuuuwxyyzzz------";
+  const p = new RegExp(a.split("").join("|"), "g");
+
+  return string
+    .toString()
+    .toLowerCase()
+    .replace(/\s+/g, "-") // Replace spaces with -
+    .replace(p, c => b.charAt(a.indexOf(c))) // Replace special characters
+    .replace(/&/g, "-and-") // Replace & with 'and'
+    .replace(/[^\w\-]+/g, "") // Remove all non-word characters
+    .replace(/\-\-+/g, "-") // Replace multiple - with single -
+    .replace(/^-+/, "") // Trim - from start of text
+    .replace(/-+$/, ""); // Trim - from end of text
+}
+
+export function getUid(name: string) {
+  const randomHex = () => Math.floor(Math.random() * 16777215).toString(16);
+  const fileNameSlug = slugify(name);
+  return randomHex() + "-" + fileNameSlug;
+}
+
+export function AvatarUpload({
+  user,
+  setSuccess,
+  setError,
+}: {
+  user: ProfileSettingsForm_UserFragment;
+  setSuccess: React.Dispatch<React.SetStateAction<boolean>>;
+  setError: (error: Error | ApolloError | null) => void;
+}) {
+  const [changeAvatar] = useChangeAvatarMutation();
+  const [fileList, setFileList] = useState<any>(
+    user && user.avatarUrl
+      ? [
+          {
+            uid: "-1",
+            name: "avatar",
+            type: "image",
+            size: 1,
+            url: user.avatarUrl,
+          },
+        ]
+      : null
+  );
+
+  useEffect(() => {
+    if (user) {
+      const avatar = user.avatarUrl;
+      if (avatar) {
+        setFileList([
+          {
+            uid: "-1",
+            name: "avatar",
+            type: "image",
+            size: 1,
+            url: avatar,
+          },
+        ]);
+      } else {
+        setFileList(null);
+      }
+    }
+  }, [user, user.avatarUrl]);
+
+  // const onChange = (info: UploadChangeParam) => {
+  //   console.log(info);
+  //   setFileList([...fileList]);
+  // };
+
+  const beforeUpload = (file: any) => {
+    const fileName = file.name.split(".")[0];
+    const fileType = file.name.split(".")[1];
+    file.uid = getUid(fileName) + "." + fileType;
+    const isJpgOrPng = file.type === "image/jpeg" || file.type === "image/png";
+    if (!isJpgOrPng) {
+      message.error("You can only upload JPG or PNG images!");
```

Comment:
Why this restriction?

Reply:
Just to restrict the file type to an image format. What other image formats should be supported?

As far as I can tell, you can't restrict what type of file is uploaded with a pre signed url.  
You can give it a content type, but it will still accept any.  
Same with file size. So, thats a problem if the user want to be malicious.  
I think a different method would need to be used to enforce file size and type on the S3 end.

Reply:
I don't see anything wrong with accepting all image types: that is, all MIME types that begin with `image/`. See [the MDN article on MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#Discrete_types) to see some examples.

Reply:
It should probably be an image format you can render in most browsers unless you're using an avatar resizing helper. You might also want to restrict things like people using GIF avatars because it can be pretty irritating...